### ISSUE TITLE
Token Authentication -  Unexpected error :Can not convert Object to String.; Code: 50000

### ISSUE THREAD

**Issue by engenb** [ably/ably-dotnet#110]
I'm following the [Token Auth tutorial](https://www.ably.io/tutorials/token-authentication#lang-java) and having some trouble.

I have implemented my api as such:
```csharp
[HttpGet]
public async Task<IActionResult> AuthGet()
{
    var rest = new AblyRest("{api key goes here}");
    var tokenParams = TokenParams.WithDefaultsApplied();
    tokenParams.ClientId = Guid.NewGuid().ToString();
    var tokenRequest = await rest.Auth.CreateTokenRequestAsync(tokenParams);
    return Ok(tokenRequest);
}
```
and setting up my client as such:
```csharp
Log.Logger = new LoggerConfiguration()
    .WriteTo
    .LiterateConsole()
    .CreateLogger();

var ably = new AblyRealtime(new ClientOptions
{
    AuthUrl = new Uri("http://localhost:5000/auth"),
    LogHander = new AblyLogger()
});

ably.Connection.On(change => Log.Logger.Information(
    change.HasError
        ? "Connection state change {previousState} => {currentState}: {error}"
        : "Connection state change {previousState} => {currentState}",
    change.Previous,
    change.Current,
    change.Reason));

var channel = ably.Channels.Get("test");
channel.On(change =>
{
    Log.Logger.Information(
        change.Error != null
            ? "Channel state change {previousState} => {currentState}: {changeReason}"
            : "Channel state change {previousState} => {currentState}",
        change.Previous,
        change.Current,
        change.Error);
});

channel.Subscribe(message =>
{
    if (message.ClientId != ably.ClientId)
    {
        Console.WriteLine(message.Data);
    }
});

while (true)
{
    var msg = Console.ReadLine();
    channel.Publish(new Message
        {
            Data = msg
        },
        (success, info) =>
        {
            Log.Logger.Debug(info.ToString());
        });
}
```
I can confirm that the client is hitting my api.  However, it's reporting an error ```Can not convert Object to String```.  My first thought was this was perhaps the Content-Type on my response was application/text and the sdk was interpreting the response as a string instead of a json object.  I have confirmed my response Content-Type is application/json.

example response from my api:
```javascript
Content-Type: application/json; charset=utf-8
Date: Wed, 21 Jun 2017 19:49:16 GMT
Server: Kestrel
Transfer-Encoding: chunked
{
    "keyName": "Oz7RFA.puyghg",
    "ttl": "01:00:00",
    "capability": {
        "resources": [
            {
                "name": "*",
                "allowedOperations": [
                    "*"
                ],
                "allowsAll": true
            }
        ]
    },
    "clientId": "7a25852e-9708-4864-a12a-5820199dfcd5",
    "timestamp": "2017-06-21T19:48:52.2411818+00:00",
    "nonce": "fa3cc548dda54f319ac7d736030abf7c",
    "mac": "OhpF6MjBR17hpX+4VhjEi3WMQcJLXQCdQ36EN2rUYNo="
}
```

I dug through [AblyAuth.cs](https://github.com/ably/ably-dotnet/blob/master/src/IO.Ably.Shared/AblyAuth.cs#L181) and, based on what I'm seeing there, the response should be getting converted properly into a TokenRequest.

Any insight into what is going wrong here would be greatly appreciated as I work through my evaluation of Ably as a messaging platform - thanks!

More detailed log output from what I described above:
```
[15:34:27 INF] Connection state change Initialized => Connecting
[15:34:27 INF] Channel state change Initialized => Attaching
[15:34:27 ERR] 08:34:27.403 #3 Error while creating transport!. Message: Can not convert Object to String.
[15:34:27 ERR] 08:34:27.408 #3 Error attaching to context Error code: 50000
Status code:
Reason: Unexpected error :Can not convert Object to String.
Reason: Unexpected error :Can not convert Object to String.; Code: 50000
   at IO.Ably.Transport.ConnectionManager.<CreateTransport>d__52.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IO.Ably.Transport.States.Connection.ConnectionConnectingState.<OnAttachToContext>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IO.Ably.Transport.ConnectionManager.<>c__DisplayClass51_0.<<SetState>b__0>d.MoveNext()
Inner exception:
Can not convert Object to String.
   at Newtonsoft.Json.Linq.JToken.op_Explicit(JToken value)
   at IO.Ably.Capability.GetResource(JProperty child)
   at IO.Ably.Capability..ctor(String capabilityString)
   at IO.Ably.CustomSerialisers.CapabilityJsonConverter.ReadJson(JsonReader reader, Type objectType, Object existingValue, JsonSerializer serializer)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.DeserializeConvertable(JsonConverter converter, JsonReader reader, Type objectType, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at IO.Ably.JsonHelper.DeserializeObject(String value, Type type)
   at IO.Ably.JsonHelper.Deserialize[T](String json)
   at IO.Ably.AblyAuth.<RequestTokenAsync>d__40.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IO.Ably.AblyAuth.<AuthoriseAsync>d__44.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IO.Ably.AblyAuth.<GetCurrentValidTokenAndRenewIfNecessaryAsync>d__39.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IO.Ably.Transport.TransportParams.<Create>d__49.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IO.Ably.Transport.ConnectionManager.<CreateTransportParameters>d__71.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IO.Ably.Transport.ConnectionManager.<CreateTransport>d__52.MoveNext()


[15:34:27 INF] Channel state change Attaching => Failed: Reason: Unexpected error :Can not convert Object to String.; Code: 50000
[15:34:27 INF] Connection state change Connecting => Failed: Reason: Unexpected error :Can not convert Object to String.; Code: 50000
[15:34:37 INF] Channel state change Failed => Failed: Reason: Channel didn't attach within the default timeout; Code: 50000
```

**Comment by engenb** [comment]
just a suggestion upon further digging into the source of the error...

perhaps consider using
```csharp
JsonConvert.PopulateObject(capabilityString, this);
```
in the Capability ctor instead of messing around with interpreting a JObject and its child JProperties

doc reference:
http://www.newtonsoft.com/json/help/html/PopulateObject.htm

**Comment by marto83** [comment]
It is something around parsing the capability string. I'll have a look at it first thing in the morning. 

There is a custom converter for capability which should handle it.

**Comment by marto83** [comment]
Looking at the response it seems that .net core MVC is serialising the capability string wrong. I'll give it a try and update your sample.

**Comment by engenb** [comment]
FWIW I fixed the issue by updating AblyAuth line 194 to
```csharp
postData = JsonConvert.DeserializeObject<TokenRequest>(signedData);
```
I can submit this via PR if you'd like.

I noticed other issues like mvc serializes the DateTime to an ISO 8601 string instead of the ticks I think the sdk is expecting.  Newtonsoft.Json will handle ticks vs. various date format strings for you, however

**Comment by marto83** [comment]
Nice. That will work for .net posting token requests but it won't work if
you have a different server side library.

It will probably help to add a helper method to serialise the token request
using the internal settings which will handle the ticks and capability
correctly.

**Comment by marto83** [comment]
Have a look. There is an internal class called jsonhelper that has custom settings. 

Initially I used to update the global serialisation settings but had to change it as wasn't fair to the other code :)

**Comment by engenb** [comment]
yup, I'm getting all up in your code over here :P

I think all you need is a [JsonConverter(...)] attribute on the TokenRequest class to convert timestamp and ttl
to the expected format.  that may also include adding a [JsonConverter(...)] to the Capability property since the expected format is somewhat non-standard for .net

or yes, you could add a .ToJson to TokenRequest to handle all the special conversion internally.  The caveat there is it requires poor ignorant devs (like me) to know they need to use that special method.  (my preference is the former option because it's handled implicitly ;))

**Comment by marto83** [comment]
Agree. It will be ideal if it is obvious. I didn't want to depend on adding the json converter properties because when I do ILMerge and merge newtonsoft.json into the ably.dll (so it doesn't cause json.net version mismatches) all these attributes became internal. 

How about if I change the signature of the method to return Task<string> which will be an already serialised version of the TokenRequest object. I'll have to call it something different and deprecate the current call

**Comment by engenb** [comment]
oh you're ILMerging in Json.Net to avoid version conflicts - that's reasonable.  well, yet another approach could be to just set your Newtonsoft.Json version all the way back to a minimum of 6.0.1.  If anyone is on a version older than that, published in 2014.... they should just update ;)

I suppose it all comes down to what your customers need.

FWIW I'm very impressed with Ably vs. other realtime vendors I've been evaluating, and I really appreciate your responsiveness @marto83!

**Comment by marto83** [comment]
You are very welcome. I'll chat to @mattheworiordan and think the best approach will be to return the tokenrequest as a serialised string. The only reason to request on is to pass it down to another client.

Thanks for taking the time for the PR :)

**Comment by marto83** [comment]
@engenb this is now fixed in the latest beta 0.8.5-beta14. I'll close the pull request. Thanks again for taking the time.
